fossas/fossa-cli
# Gradle: Adds resolution api tactic for analysis

## Overview

When analyzing android gradle builds, which are multi project, and multi-module with android variants, we are hitting:

```
org.gradle.internal.component.AmbiguousVariantSelectionException: The consumer was configured to find an API of a component, preferably optimized for Android, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'release', attribute 'partner' with value 'xxx', attribute 'org.jetbrains.kotlin.platform.type' with value 'androidJvm'. However we cannot choose between the following variants of project :common:yyy-xxx:
  - Configuration ':common:yyy-xxx:releaseApiElements' variant android-aidl declares an API of a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'release', attribute 'org.jetbrains.kotlin.platform.type' with value 'androidJvm':
      - Unmatched attributes:
          - Provides attribute 'artifactType' with value 'android-aidl' but the consumer didn't ask for it
          - Provides attribute 'com.android.build.gradle.internal.attributes.VariantAttr' with value 'release' but the consumer didn't ask for it
          - Doesn't say anything about its target Java environment (preferred optimized for Android)
          - Doesn't say anything about partner (required 'xxx')
```

This PR, changes:

- Gradle `jsondeps.gradle` script to include results from using gradle's resolution result api
- Builds the graph using gradle's resolution result api

Prerequisite PR: https://github.com/fossas/fossa-cli/pull/720

## Acceptance criteria

- Gradle: CLI can analyze multi module android project.

## Testing plan

**Test cases:**

Android Multi-Module (with Library) Test:

- [x] https://github.com/HaenaraShin/Android-Multi-Module-Sample.git
- [x] https://github.com/JeroenMols/ModularizationExample.git 

Regression Test:

- [x] https://github.com/gradle/gradle-build-scan-quickstart.git
- [x] https://github.com/danieldisu/GradleKotlinStarter.git
- [x] https://github.com/gwonsungjun/gradle-multi-module.git)


**To test,**

1. Perform fossa analysis: `fossa analyze -o | jq`. Ensure results are reported.
2. Compare result of (1) against, `.\gradlew build --scan` results. Number of dependencies, and dependencies edges should be exactly the same.

## Risks

This changes default gradle tactic, so there is high blast zone if there is edge case unaccounted for. 

## Open Question

~Should we introduce this change behind experimental configuration flag? Instead of drastically changing existing tactic?~

We will replace existing tactic.

## References

Closes: https://github.com/fossas/team-analysis/issues/817

## Checklist
